mu / Sigma (scale-independent) modified gravity support

c-d-leonard opened the pull request:
This PR implements support for the scale-independent version of the mu / Sigma parameterisation of modifications to GR (for a definition of these functions see, for example, Simpson et al. 2013, the CFHTLenS modified gravity constraints paper). 

At the moment this is not quite ready for review but it is close. Here is what is still missing:
- benchmarks for growth factor, P(k), and Cl's (but test code of the C library is there waiting)
- python tests
- updates to existing example scripts and notebooks.

If you can contribute benchmarks please let me know (@TessaBaker perhaps?)

c-d-leonard commented:
Okay, I've now merged master into this branch following all the v1 updates and the C and python tests are passing. Next up is to integrate the benchmarks (right now the MG tests just test the interface or are placeholders).

mishakb commented:
@c-d-leonard  Just following up that I submitted comments for the review last week (just in case they fell under the stack) :)

c-d-leonard commented:
@mishakb Thank you for the nudge! I will make it a priority to address these by early next week.

c-d-leonard commented:
@mishakb Just to update you: so, I've merged master into this branch and got it working again. To address your above questions:
- Good catch about CMB lensing. I am pretty sure I did at some point have the appropriate Sigma factors in the CMB lensing integral, however the cl code has been seriously altered since then and I think it's gotten lost. That said, I've just been staring at ccl_cls.c for more than an hour and I can't understand where the CMB lensing integral is calculated (except in the limber case where I can track it). Maybe @beckermr or @damonge could chime in pointing me to where this integral is specified for the non-limber case?
- The number counts tracer is affected by the mg parameters in the following ways:
1) the power spectrum which feeds in to it is modified by mu0 (implemented in ccl_power.c)
2) the magnification effect is altered by Sigma, you can see this in the function clt_init_wM.
I tried to add a few comments to clarify this.
- Yes, the growth rate should be implemented in the way you say following our previous discussions.

mishakb commented:
@c-d-leonard thanks for sending the update. I will keep an eye open on how the cmb lensing will be addressed.  For the growth, I will check if such a redshift/time "switch" is there. I don't recall coming across it in the code but will check again.

c-d-leonard commented:
@mishakb re your comment on redshift / time dependency of growth - I think I totally misunderstood your question the first time. I thought you were asking about whether we normalized the growth rate to GR at early times, but now I think you were asking about the assumed redshift dependence of mu. Is the second one correct?

Assuming yes, I'll clarify: currently this PR assumes that mu(z) = mu0 * OmegaL(z) / OmegaL_0 and similarly for Sigma. I think we said that someone in your group was interested in doing another PR after this one gets merged to generalize this. Does that answer your question?

damonge commented:
@c-d-leonard : CMB (and non-CMB) lensing are not implemented in the non-Limber case. Does that make sense?

c-d-leonard commented:
@damonge oh okay, that explains why I couldn't find where it was implemented :) Has that always been true? Is it something we think is going to change in the future or are we happy with limber-only for those cases?

c-d-leonard commented:
@mishakb, given @damonge's reply above I will just go ahead and make sure that the appropriate factors are implemented for CMB lensing in the Limber case.

mishakb commented:
@c-d-leonard Good for CMB-Lensing then (@damonge ). I checked again for clt_init_wM and fine too. For the early times recovery of GR, you got my question right the first time.  So by the parametrization above, at early times Omega_L(z) becomes smaller and smaller until vanishing so mu=0 and GR is recovered as well as it growth at early times. So this addresses this question.

mishakb commented:
@c-d-leonard incidently, I was not able to see your new comments about the K-g correlation window?

c-d-leonard commented:
> @c-d-leonard incidently, I was not able to see your new comments about the K-g correlation window?

@mishakb The way it's set up, any correlation is made up of two tracers; here it would be one lensing-type tracer (for K) and one number counts tracer (for g). The MG effect on the K tracer is through the modified power spectrum and through the factors of Sigma in clt_init_wL (around like 422 of ccl_cls.c). The MG effect of the number counts tracer is through the modified matter power spectrum alone (unless there is magnification); this comes in either in transfer_nc (around line 723) for the limber case or within CCL/build/ANGPOW/src/ANGPOW/inc/Angpow/angpow_ccl.h in the angpow case.

mishakb commented:
@c-d-leonard sorry for not making my question clear enough. I understand and agree with your comment above and that is indeed how auto-cross-correlations are implements.  What I really wanted to mention is that I did not see the comments in the code itself and recall that you said/wrote that you added some. (maybe I am just missing them).

c-d-leonard commented:
> @c-d-leonard it would be great to put an "MG" label or "flag" everywhere where you made changes. I think there are already at almost each place so just to make sure none are missing.

@mishakb I can do this, but I'm not sure if we actually want to. I understand that it would be useful for reviewing, but you can imagine that if we had different flags for every PR we merged into master, the code in master would get pretty messy pretty fast. Maybe @elisachisari can comment on this stylistic issue?

c-d-leonard commented:
> @c-d-leonard sorry for not making my question clear enough. I understand and agree with your comment above and that is indeed how auto-cross-correlations are implements. What I really wanted to mention is that I did not see the comments in the code itself and recall that you said/wrote that you added some. (maybe I am just missing them).

Right, I added some at e.g. lines 739 and 952.

mishakb commented:
> @c-d-leonard it would be great to put an "MG" label or "flag" everywhere where you made changes. I think there are already at almost each place so just to make sure none are missing.

@mishakb I can do this, but I'm not sure if we actually want to. I understand that it would be useful for reviewing, but you can imagine that if we had different flags for every PR we merged into master, the code in master would get pretty messy pretty fast. Maybe @elisachisari can comment on this stylistic issue?

@c-d-leonard I see your point and it probably makes sense. So perhaps just the comments are good enough to trace the changes plus the git of course. @elisachisari 

@c-d-leonard let me know when you are done with all changes.

c-d-leonard commented:
@mishakb I think it's ready to look at - I am just fixing some issues with the Travis build but I don't have any more substantial changes planned at the moment.

What's the status on understanding at which scales we expect what agreement with benchmarks?

c-d-leonard commented on benchmarks/ccl_test_power_nu.c:
```diff
@@ -19,6 +19,8 @@ CTEST_DATA(power_nu) {
   double Omega_k;
   double w_0[3];
   double w_a[3];
+  double mu_0;
```
Good catch, I forgot to initialize these variables. Fixing.

c-d-leonard commented on benchmarks/ccl_test_sigmaM.c:
```diff
@@ -19,6 +19,8 @@ CTEST_DATA(sigmam) {
   double Omega_k[5];
   double w_0[5];
   double w_a[5];
+  double mu_0;
```
Similarly, fixing.

c-d-leonard commented on doc/0000-ccl_note/main.tex:
```diff
@@ -568,6 +583,18 @@ \subsubsection{Impact of baryons}
+\subsubsection{Modified gravity ($\mu, \Sigma$)}
+\ccl supports the quasistatic parameterization of modified gravity with scale-independent deviations from GR which arise at late times (currently with functional form given in equation \ref{muSigform}). Under these conditions, the matter power spectrum in a modified gravity scenario can be computed simply by re-scaling the GR matter power spectrum with the modified gravity growth factor.
+
+For each power spectrum calculation method above, we first first compute the power spectrum in GR at a grid in $a$ and $k$ as usual, in preparation for producing splines in these variables. If ${\tt mu\_0} \ne 0$, we then take the ratio of the growth factor $D$ at the given value of ${\tt mu\_0}$ to the GR (${\tt mu\_0}=0$) growth factor, and multiply the GR power spectrum by this ratio squared:
```
fixing

c-d-leonard commented on doc/0000-ccl_note/authors.tex:
```diff
@@ -61,7 +61,7 @@
 \author{An\v{z}e Slosar}
 \affiliation{Brookhaven National Laboratory, Physics Department, Upton, NY 11973, USA}
-\author{Tilman Tr\"oster}
+\author{Tilman Tr\""oster}
```
It does not ... I have no idea what's up here. @tilmantroester would you be able to fix this? Sorry for messing it up, I can't figure out how to fix it.

tilmantroester replied:
This should be a single double quote, i.e., Tr\"oster.

elisachisari replied:
This must have happened when you compiled the note again. authors.tex gets automatically generated from authors.csv. I would discard this change.

c-d-leonard replied:
I am trying to discard it, and I believe I have authors.csv back exactly the same as in master, but when I compile on my machine I still get messed-up output. Could somebody else try compiling to note from this branch on their computer?

c-d-leonard commented on pyccl/background.py:
```diff
@@ -222,3 +223,30 @@ def rho_x(cosmo, a, species, is_comoving=False):
+        cosmo (:obj:`ccl.cosmology`): Cosmological parameters.
+        a (float or array_like): Scale factor(s), normalized to 1 today.
+    Returns:
+        Sig_MG (float or array_like): Modification to Poisson equation
```
I added a comment pointing out the equation in the CCL note, is that okay or do you want an external reference?

elisachisari replied:
Could we have an external reference? Eq. numbers in the CCL note might change as text evolves.

c-d-leonard replied:
Sure, added.

c-d-leonard commented on pyccl/background.py:
```diff
@@ -222,3 +223,30 @@ def rho_x(cosmo, a, species, is_comoving=False):
+        cosmo (:obj:`ccl.cosmology`): Cosmological parameters.
+        a (float or array_like): Scale factor(s), normalized to 1 today.
+    Returns:
+        mu_MG (float or array_like): Modification to Poisson equation
```
Same as above, I added a reference to the note.

c-d-leonard replied:
Switched to an external reference as above.

c-d-leonard commented on pyccl/core.py:
```diff
@@ -1,6 +1,7 @@
 """The core functionality of ccl, including the core data types. This includes
 the cosmology and parameters objects used to instantiate a model from which one
 can compute a set of theoretical predictions.
+can compute a set of theoretical predictions.
```
good catch, removed

c-d-leonard commented on benchmarks/ccl_test_power_MG.c:
```diff
@@ -0,0 +1,218 @@
+  double z=0.; //Other redshift checks are possible 
+  int stat=0;
+
+	for(i=0;i<nk;i++) {      
```
done

c-d-leonard commented on src/ccl_correlation.c:
```diff
@@ -462,9 +469,17 @@ void ccl_correlation_3d(ccl_cosmology *cosmo, double a,
+  //    }
+  //} else {
+	  for (i=0; i<N_ARR; i++){
+          pk_arr[i] = ccl_nonlin_matter_power(cosmo, k_arr[i], a, status);
```
That's right, that commented-out code was left-over from before we had a good set-up for catching this issue at the power spectrum level. So now, if you are asking for the correlation function (or any other nonlinear power spectrum based quantity) in modified gravity, we are requiring that matter_power_spectrum be set to linear. This is implemented in ccl_power.c. I removed the commented out code.

c-d-leonard commented on src/ccl_correlation.c:
```diff
@@ -444,6 +444,13 @@ void ccl_correlation_3d(ccl_cosmology *cosmo, double a,
+  // Print a message informing the user that if they are using mu / Sigma
+  // parameterisation, Cl's will be computed using the linear power spectrum
+  // because we do not have support for nonlinearity in this parameterisation
+  //if ( fabs(cosmo->params.mu_0)>1e-15 || fabs(cosmo->params.sigma_0)>1e-15 ){
```
Same reason as above; this was implemented before we decided to check for this at the power spectrum level. I removed it.

c-d-leonard commented on benchmarks/ccl_test_growth_lowz.c:
```diff
@@ -46,6 +48,24 @@ static void read_growth_test_file(double z[6], double gf[5][6])
   }
   fclose(f);
+  
+  /*// Now get the growth in the models with non-zero mu_0 and sigma_0
```
Good point. A lot of these commented out sections can be explained by the fact that I wrote the code to incorporate benchmark comparisons for these new features before we actually had benchmarks. Here, I was hoping to get some benchmarks for the growth rate, but I haven't got anything I can use for that (Tessa had given me some but they have different redshift dependence for the modified gravity functions). I am removing this code.

c-d-leonard replied:
Let me know if you feel strongly we do need this kind of a benchmark and I can ask Tessa to modify her code.

elisachisari replied:
Are these cases tested at all through the C_ell test? Or is that assessing a different aspect of the model? I'd like to understand what current "gaps" we have in the benchmarking.

elisachisari replied:
We discussed this and these tests would be covered by the power spectrum tests.

c-d-leonard commented on benchmarks/ccl_test_growth_lowz.c:
```diff
@@ -85,8 +109,10 @@ static void compare_growth(int model, struct growth_lowz_data * data)
-  ccl_parameters params = ccl_parameters_create(data->Omega_c, data->Omega_b, data->Omega_k[model], data->Neff, data->mnu, data->mnu_type, data->w_0[model], data->w_a[model], data->h, data->A_s, data->n_s,-1,-1,-1,-1,NULL,NULL, &status);
-  params.Omega_g=0;//enforce no radiation
+  ccl_parameters params = ccl_parameters_create(data->Omega_c, data->Omega_b, data->Omega_k[model], data->Neff, data->mnu, data->mnu_type, data->w_0[model], data->w_a[model], data->h, data->A_s, data->n_s,-1,-1,-1,data->mu_0[model], data->sigma_0[model],-1,NULL,NULL, &status);
+  if (model<=4){   // Don't know yet whether this will be true for MG benchmarks.
```
Removed because as noted above we don't actually have benchmarks for the growth rate.

c-d-leonard commented on benchmarks/ccl_test_growth_lowz.c:
```diff
@@ -178,6 +203,29 @@ CTEST2(growth_lowz, model_5) {
 }
 
-CTEST2(growth_lowz, mgrowth) {
+/*CTEST2(growth_lowz, model_6) {
```
See above.

c-d-leonard commented on benchmarks/ccl_test_power_MG.c:
```diff
@@ -0,0 +1,218 @@
+  ccl_cosmology_free(cosmo);
+}
+
+/*
```
Good question - the last four tests are commented out because of the ongoing issues with understanding how well we expect our code to match the benchmark produced by the UT group at larger k. These I expect we will re-include before we merge with appropriate tolerance for discrepancy at appropriate k (the UT group is actively working on figuring this out at I understand). The first four tests, however, were just commented out in error; I included them again.

c-d-leonard commented on src/ccl_cls.c:
```diff
@@ -364,8 +366,14 @@ static void clt_init_wM(CCL_ClTracer *clt,ccl_cosmology *cosmo,
+      // If mu / Sigma parameterisation of modified gravity is in effect,
+	  // add appropriate factors of Sigma before splining:
+      if ( fabs(cosmo->params.sigma_0) ){
+		    y[j] = y[j] * (1. + ccl_Sig_MG(cosmo,ccl_scale_factor_of_chi(cosmo,x[j], status), status));
```
I added a status check here.

c-d-leonard commented on src/ccl_cls.c:
```diff
@@ -442,8 +452,14 @@ static void clt_init_wL(CCL_ClTracer *clt,ccl_cosmology *cosmo,
+      // If mu / Sigma parameterisation of modified gravity is in effect,
+	  // add appropriate factors of Sigma before splining:
+      if ( fabs(cosmo->params.sigma_0) ){
+		    y[j] = y[j] * (1. + ccl_Sig_MG(cosmo,ccl_scale_factor_of_chi(cosmo,x[j], status), status));
```
Added

c-d-leonard commented on src/ccl_cls.c:
```diff
@@ -793,6 +812,11 @@ static double transfer_cmblens(int l,double k,ccl_cosmology *cosmo,CCL_ClTracer
+    // If muSigma parameterisation of gravity is in effect and 
+    // Sigma0>0, add the relevant factor here.
+    if (fabs(cosmo->params.sigma_0)>1e-15){
+        w = w * (1. + ccl_Sig_MG(cosmo,ccl_scale_factor_of_chi(cosmo,chi, status), status)); 
```
Added.

c-d-leonard commented on tests/ccl_test_cls_MG.c:
```diff
@@ -0,0 +1,276 @@
+    if(l<=0)
+      ell_correct=1;
+    else
+      ell_correct=l*(l+1.)/sqrt((l+2.)*(l+1.)*l*(l-1.));
```
Right, so again this is a case where originally there were going to be Cell benchmarks for this case, and now it looks like there will not be, but I wrote this code before I really knew. We will now have real space two point functions instead. I will remove this file from the repo; I need to add one which checks the correlation function benchmarks we have from Cristhian.

c-d-leonard commented:
@elisachisari Thanks so much for the review! I addressed your comments one by one above but in general:
- I need to add a tests file to include the benchmarks I got recently for real space two point correlation functions, rather than Fourier space (see comment above)
- I am very confused about what's going on with the tests. It looks like they are passing on linux and failing on mac, which makes it very hard for me to debug because I don't have a mac ... I may need some help with this.
- We still need to get the final word on the expected tolerance at a given scale of the difference in the benchmarks and CCL for the matter power spectrum, @mishakb will have more information on that status of this.

mishakb commented:
@c-d-leonard I have been working on the comparison rationales/expectations but got slowed down by the collaboration wide presentation today and the last days of teaching this week.  I will work on it again tomorrow.  Glad that @elisachisari looked into the tests/benchamarks code changes for MG as that is the part I wanted to ask about before to look into.  Are the only related directories examples and benchmarks?

mishakb commented:
pushed wrong button by mistake.  :(

elisachisari commented on doc/0000-ccl_note/contributions.tex:
```diff
@@ -27,7 +27,7 @@
 Sukhdeep Singh: Contributed to the correlation functions code. \\
 An\v{z}e Slosar: Wrote and reviewed code. \\
-Tilman Tr\"oster: Wrote code for user-changable precision parameters, added distance and growth factor tests, found and fixed bugs. \\
+Tilman Tr\""oster: Wrote code for user-changable precision parameters, added distance and growth factor tests, found and fixed bugs. \\
```
Same as above.

c-d-leonard replied:
Similarly to the above, I can't get what I think is the correct thing in authors.csv to produce the right thing in authors.tex, contribution.tex, or the pdf when I make the note. @elisachisari would you mind trying it on your computer?

elisachisari replied:
I would simply replace authors.csv and authors.tex with the master version. They shouldn't have changed in this PR.

c-d-leonard commented:
@mishakb the directories relevant for the tests are benchmarks/ and tests/

@elisachisari To update: I have started on a script to include tests against the correlation function benchmarks Cristhian sent me but (as you will have seen in the chat Mustapha) I am waiting on some information from him (e.g. the dNdz's he used), so I'm not going to push that script to the repo yet. I will hopefully be able to work on this at the beginning of next week a little but like I said I am traveling for the next two weeks so I won't have that much time.

Also the Travis build seems to be failing erratically (it's failing C tests on mac which it passes on linux, which is super weird and hard for me to debug). If someone with a mac could run the tests on this branch, that would be a help.

elisachisari commented:
The C tests are failing in my Mac:
`TEST 87/95 power_MG:MG_nonlin_error ERROR 1040: ccl_background.c: ccl_cosmology_compute_growth(): Support for the growth rate in cosmologies with massive neutrinos is not yet implemented.`
The python tests fail here:
`Test halo model and supporting functions. ... mu0=0.000000, sig0=0.000000
ok
ccl_test_pyccl_interface.test_halomod(pyccl.Cosmology(wa=0.0, n_s=0.96, Neff=3.046, sigma8=0.8, w0=-1.0, bcm_etab=0.5, bcm_log10Mc=14.079181246047625, A_s=None, h=0.67, sigma_0=0.1, mu_0=0.1, Omega_k=0.0, bcm_ks=55.0, Omega_c=0.27, Omega_b=0.0490086879038, Omega_g=None, m_nu=0.0, mnu_type=None, z_mg=None, df_mg=None, matter_power_spectrum='linear', baryons_power_spectrum='nobaryons', halo_concentration='duffy2008', transfer_function='boltzmann_class', mass_function='tinker10', emulator_neutrinos='strict'),)
Test halo model and supporting functions. ... Abort trap: 6`

The Travis build that is successful doesn't seem to be running the C tests (I can't find them in the output, which is very strange): https://travis-ci.org/LSSTDESC/CCL/jobs/530873780
The Travis build that fails, fails both for C and python: https://travis-ci.org/LSSTDESC/CCL/jobs/530873782 The C error is the same as above, but there seems to be some difference in the python failure. I don't know why the python tests aren't failing in some of the Travis builds.

When the tests run (both in C and in python), I get prints to screen of "mu0=0.0000" and "sigma=0.0000". We should remove this before merging but for now they seem useful to debug.

c-d-leonard commented on doc/0000-ccl_note/main.tex:
```diff
@@ -714,6 +742,13 @@ \subsubsection{Exact expressions}
+        \begin{equation}
+          \Delta_\ell^{\rm M}(k)=-\ell(\ell+1)\int \frac{dz}{H(z)} W^{\rm M}(z) (1 + \Sigma(z)) T_{\phi+\psi}(k,z) j_\ell(k\chi(z)),
+          \label{eq:deltaM}
+        \end{equation}
```
Good points, fixed (and fixed for the lensing and cmb lensing transfer equations as well).

c-d-leonard commented on pyccl/core.py:
```diff
@@ -207,6 +207,19 @@
+#    'GR':   lib.GR,
+#    'muSigma_MG':  lib.muSigma_MG
+# }
+"""dict: Types of modified gravity models.
```
Good catch. I put this in a long time ago when we weren't sure what the scope of this PR was. I removed it, but it will be there in the commit history if someone wants to include it again with the implementation of other modified gravity options.

c-d-leonard commented on doc/0000-ccl_note/main.tex:
```diff
@@ -740,7 +782,11 @@ \subsubsection{Exact expressions}
+with dependence on $\mu(z)$ via modifications to $T_{\phi+\psi}(k,\chi)$.
 
 \subsubsection{The Limber approximation}
 As shown above, computing each transfer function involves a radial projection (i.e. an integral over redshift or $\chi$), and thus computing full power spectrum consists of a triple integral for each $\ell$. This can be computationally intensive, but can be significantly simplified in certain regimes by using the Limber approximation, given by:
```
Good idea, done.

c-d-leonard commented on src/ccl_class.c:
```diff
@@ -489,22 +488,89 @@ void ccl_cosmology_compute_linpower_class(ccl_cosmology* cosmo, int* status) {
+      // and mu ! = 0 : get the unnormalized growth factor in MG and for 
+      // corresponding GR case, to rescale CLASS power spectrum
+    if ( fabs(cosmo->params.mu_0)>1e-14){
+	    // Set up another cosmology which is exactly the same as the 
```
Here I checked on mu_0 because only mu_0 affects the power spectrum, which is what is being calculated in this file. Similarly, in the case of the checks in ccl_cls.c, only Sigma_0 is checked because only Sigma_0 affects the lensing and magnification tracers directly (dependence on mu0 is automatically inherited from any dependence in P(k)). Does that answer this question?

mishakb replied:
I used mu and sigma for the two cases but what I really meant is the consistency of the if condition 
if fabs(cosmo->params.mu_0)>1e-14
versus 
if (fabs(cosmo->params.mu_0) 

at diffrent places

c-d-leonard replied:
Oh I see. Where are you seeing the second one? (File and line number would be helpful.)

mishakb replied:
It may have been in an older file that I did not stash when I wrote that. If you don't see it don't worry about it.

c-d-leonard commented on src/ccl_cls.c:
```diff
@@ -364,12 +366,22 @@ static void clt_init_wM(CCL_ClTracer *clt,ccl_cosmology *cosmo,
+      // If mu / Sigma parameterisation of modified gravity is in effect,
+	  // add appropriate factors of Sigma before splining:
+      if ( fabs(cosmo->params.sigma_0) ){
+		    y[j] = y[j] * (1. + ccl_Sig_MG(cosmo,ccl_scale_factor_of_chi(cosmo,x[j], status), status));
```
That's a good point. It's the note which is incorrect, so I've removed that line. The dependence on mu_0 is instead via the power spectrum in equation 31 of the note.

mishakb commented on include/ccl_config.h:
```diff
@@ -51,7 +51,13 @@ typedef enum baryons_power_spectrum_t
+/** 
+ * Modified gravity typedef.
+ * Specified what, if any, parameterisation
+ * is being used for modifications to GR
```
is this meant to be kept here?

c-d-leonard replied:
no, good point, I'm removing it.

c-d-leonard commented:
@mishakb (cc @elisachisari) - Thanks a lot for giving careful thought to the question of how well the benchmarks for P(k) should agree. Given your statement that you wouldn't expect agreement above 1e-2 and that we should only compare at subhorizon scales (which I agree with), I have made the tolerance for the MG P(k) tests 1e-2 and I have only checked the ones with MG parameters nonzero at scales smaller than k_h =5e-3. Of course, this is sort of reverse-engineered, because I checked to see at what larger scales the benchmarks started to differ from CCL at greater than 1e-2. Nevertheless I think the fact that we get this agreement is a pretty good check.

With regards to your request for the P(k)'s from CCL to which I am comparing - I'm not comparing to an output file of P(k) but rather CCL gets called and compared to the output from MGCAMB / ISiTGR actively when the tests are run. If you do want to output what is being compared, you could do that pretty easily though - you'd find it in CCL/benchmarks/ccl_test_MG.c.

I still haven't heard back from Cristhian on information I need to incorporate the correlation function benchmarks. I will ping him again.

c-d-leonard commented:
@elisachisari I have no clue what's going on with the tests. The C tests at least are passing locally for my on my laptop. Let's see how these latest commits fare with travis I guess....

beckermr commented:
Hmmmmm. I am looking at the tests. A few things

1. The linux tests for the C code are not actually running! My first clue was the short testing time. Our test suite takes ~40 minutes and it finished in ~25. See here https://travis-ci.org/LSSTDESC/CCL/jobs/531997653#L2311

2. For the OSX tests, one of the failures looks like something in the C test suite. 

>>> TEST 91/99 power_MG:MG_pk_model3 ERROR 1040: ccl_background.c: ccl_cosmology_compute_growth(): Support for the growth rate in cosmologies with massive neutrinos is not yet implemented.

I don't understand this one, but presumably, there is a section of code that is calling the growth function with a neutrino model. This will need to be tracked down in the file. I am guessing we assumed this was working due to the linux tests but actually it is not since those did not run. 

3. For the OSX python tests, I don't know at all what is going on. 

Short of digging into the code itself, I think you will need to run them all locally and investigate in detail. If you cannot get them to fail locally, try clearing out the build dir etc or use a fresh installation.

c-d-leonard commented:
Thanks @beckermr and sorry about the slow reply (I've been travelling). I did a new clone and build and still can't get the C checks to fail. So, I think I'm going to have to proceed with the world's slowest debugging strategy and debug via Travis.

c-d-leonard commented:
@elisachisari @beckermr @mishakb The Travis tests are passing now! What was happening was that there was a bug introduced in getting the neutrinos masses in setting up a temporary equivalent GR cosmology object; only one of the three masses were getting set. Locally on my laptop the other masses were also by default being initialized to zero for some reason, but on the Travis build they were being set to nonsense non-zero values.

I think now the only thing is to get the correlation benchmarks included. I'm still going back and forth with Cristhian on understanding how he produced those so I can do the right comparison.

mishakb commented:
@cdleonard:  We discussed yesterday the problem here at UTD and we compared ISITGR, MGCAMB (and CAMB) using GR limit but also for the case MG params=0.1 and all checks just fine with mostly around 10^{-3} differences. Therefore, the difference has to be in the way that we have changed in ISiTGR/CAMB/MGCAMB the lensing from DES bins and  specs (as implemented in CAMB/ISiTGR/MGCAMB) to those of the CCL code and benchmarks. We are doing some further checks and will be in touch.

mishakb commented:
Attached and below are the results of comparison from @gqcristhian and I for the for the comparisons between CCL and ISiTGR (CAMB) for the GR and MG cases for the 4 correlation functions: 

So basically the relative errors  are :

w(theta)  ------> the relative error is less than 1.8x10^-2 in all cases

gammat  ------> the relative error is less than 6x10^-2 in all cases

xim  ------------> the relative error is less than 6.8x10^-2 in all cases

xip  -------------> the relative error is less than 1.6x^-2 in GR, but less than 2.5x10^-1 in MG cases due some points.  

Note also that for xip in bin(1,2) for GR and MG the last value of theta causes a large relative error for some reason. The correlation functions with the CCL data, provides in total, 180 theory predictions points. Of these 180 points, 179 have a reasonable relative error. However, 1 data point corresponding to xip, in bin (1,2), for the last value of theta (we have 15 values for theta, just the highest one) is different from CCL.  Since this issue is present in GR case as well, I am not sure if this is worth further consideration but we would like to hear what you think. 

[plotsMG.zip](https://github.com/LSSTDESC/CCL/files/3375448/plotsMG.zip)
[plotsGR.zip](https://github.com/LSSTDESC/CCL/files/3375452/plotsGR.zip)

mishakb commented:
Here are the plot for xip where we removed the last point of integration.  For all cases xip relative error is less than 2.1x10^-2

[NewPlots.zip](https://github.com/LSSTDESC/CCL/files/3375792/NewPlots.zip)

mishakb commented:
As agreed at telecon, here is a brief summary of exchanges or files related to this PR but have been done via slack-channel:

Below is part of an initial file for possible implementations of MG parameters in CCL prepared in early 2018 by Weikang Lin and Mustapha Ishak that was discussed with Danielle Leonard in May 2018.  The Method I below was adapted/completed and implemented (see CCL note):  

Method I. Some general steps for quasi-static limit :
-	Growth function (and growth rate). The file is in src/ccl_background.c. 
-	The linear matter power spectrum is obtained by multiplying the growth factor. The growth factor is affected by μ. 
-	If the matter power spectrum is scaled and stored in the spline, the total growth factor needs to be normalized to 1 today and ∝ a in the distant past as usual.
-	The total growth factor (and growth rate) should be used in the RSD transfer function. Just need to make sure the total one is being used in it.
-	Three “transfer functions” (weak lensing, magnification and CMB lens). The file is in scr/ccl_cl.c. They are affected by Σ via a modified Poisson equation for relativistic particles. 1/k^2—>Σ/k^2. 
-	Declare any additional parameters in include/ccl_core.h

Method II. Another way and σ_R. 
-	Another way is not to scale the matter power spectrum in the spline, but use whatever CLASS gives.
-	The normalization in this way will be different. The unmodified growth function is normalized to 1 today and ∝ a in the distant past. The modified growth (additional part) should be 0 in the distant past, which means the total growth function is not normalized to 1 today.
-	If someone needs to output the matter power spectrum, multiply the output with a modified growth factor. 
-	In this way, additionally one needs to change σ_R to include the modified growth factor today. The file is in scr/power.c. This can be done by multiplying the sigmaR_integrand by a growth factor (scale dependent or not). Doesn’t seem to have a σ_R(z) at a arbitrary redshift, so we don’t need to worry about that.


February 2019 – March 2019: a number of discussions and file exchanges between Danielle, Cristhian, Weikang and Mustapha on slack channel to get the MG matter power spectrum P(k) from CCL to agree with that from ISiTGR/CAMB. The implementation of MG in CCL is different than ISiTGR. In CCL the matter power spectrum of GR is rescaled using the ratio of growth factors MG/GR while in ISITGR the potentials are modified to MG and the power spectrum calculated accordingly. Results reported in benchmarks and expected and found to be at the 10^{-2} level. Results available in the benchmarks.

April 2019 -- July 9th 2019: a number of discussion and file exchanges between Danielle, Mustapha and Cristhian on slack channel to get the MG 3 X 2pts power spectra and correlations to agree between CCL and ISiTGR.  Different integration methods and configuration of test data required some good deal of work. This included the implementation of common redshift bins and distributions and other data settings. After some intense debugging and fixing of integration limits and some conventional pre-factors for the cls and correlations, an agreement between the correlation between CCL and ISiTGR of ~1-5 X 10^{-2} was reached, in accordance with the expectation. Results available in the benchmarks.

c-d-leonard commented:
@elisachisari @mishakb I just added the benchmark files to the test script and the tests are passing! locally anyway, let's hope the travis build agrees. I did change the factor by which we multiply the forecast errors to be 0.2 for both fftlog and bessel (before it was 0.1 for bessel). This still means that the discrepancy is less than 20% of the forecast error on the correlation functions, which is not too bad at all. Are we okay with this?

I still need to add a comment to the note stating not to use the power spectrum on large scales outside of the quasistatic regime with mu / Sigma != 0. I will probably end up pushing that tomorrow but I think code wise this is ready for a final look over.

c-d-leonard commented:
Oh also I renamed the benchmark files because the names were long enough that they were causing the test script to fail flake8 tests for python formatting, and I removed the files not actually used to keep the data repository cleanish.

elisachisari commented on benchmarks/data/CCL_cosmoPar.dat:
```diff
@@ -0,0 +1,15 @@
+ H0   70.001831054687500     
```
What is this file used for?

c-d-leonard replied:
nothing, it was a file pushed to the repo with the benchmarks but we're not using it, I'll remove it.

elisachisari commented on benchmarks/data/items_ISiTGR_gammat_mu_0p1_Sigma_0.dat:
```diff
@@ -0,0 +1,60 @@
+           3           1           1           1
```
What does this file do?

c-d-leonard replied:
same as above.

elisachisari commented on benchmarks/data/items_ISiTGR_gammat_mu_0p1_Sigma_0p1.dat:
```diff
@@ -0,0 +1,61 @@
+           3           1           1           1
```
Same question as previous file. What does file do?

c-d-leonard replied:
same as above.

elisachisari commented on tests/ccl_test_pyccl_interface.py:
```diff
@@ -925,6 +925,13 @@ def test_halomod():
     for cosmo_mg in reference_models_mg():
         yield check_massfunc_MG, cosmo_mg
 
+def test_haloprofile():
```
Why does this appear in this PR?

c-d-leonard replied:
For me it's not showing up as a change in this PR. It's just lines 925 - 927 above it. Do you see line 928 as a new line in this PR?

elisachisari replied:
Oh, it must have been brought in with master. I'm just viewing my changes since I last reviewed, so those show up as well.

mishakb commented:
My previous comments and suggestions have been addressed. Benchemarks for the power spectra and correlation functions for clustering, lensing and their cross-correlation against external code ISiTGR are at the expected theoretical level of agreement. These have been documented in the wiki with up to date references to ISiTGR repository. CHANGELOG and note have been updated. I approve this PR.
